nomos-services/chain-service/src/bootstrap/ibd.rs

```diff
@@ -1,50 +1,785 @@
+    async fn download_from_peer<CryptarchiaState, BlockProcessor>(
+        &self,
+        mut cryptarchia: Cryptarchia<CryptarchiaState>,
+        network_adapter: NetAdapter,
+        peer: NetAdapter::PeerId,
+        block_processor: &mut BlockProcessor,
+    ) -> Result<Cryptarchia<CryptarchiaState>, (Cryptarchia<CryptarchiaState>, IbdError)>
+    where
+        CryptarchiaState: cryptarchia_engine::CryptarchiaState + Send,
+        BlockProcessor: processor::BlockProcessor<Block = NetAdapter::Block> + Send + Sync,
+    {
+        // Track this to avoid downloading the same block multiple times.
+        let mut latest_downloaded_block: Option<HeaderId> = None;
+
+        // Repeat until the peer's tip is reached.
+        loop {
+            // Each time, set `target` to the most recent peer's tip.
+            match network_adapter.request_tip(peer).await {
+                Err(e) => {
+                    return Err((cryptarchia, IbdError::ErrorFromBlockProvider(e)));
+                }
+                Ok(tip) => {
+                    if !should_download(tip.id, tip.slot, &cryptarchia) {
+                        break;
+                    }
+                    let target = tip.id;
+
+                    // Download a single stream of blocks from the peer.
+                    (cryptarchia, latest_downloaded_block) = self
+                        .download_stream(
+                            cryptarchia,
+                            &network_adapter,
+                            peer,
+                            target,
+                            latest_downloaded_block,
+                            block_processor,
+                        )
+                        .await?;
+
+                    // Stop if the target block has been downloaded/added.
+                    if cryptarchia.consensus.branches().get(&target).is_some() {
+                        break;
+                    }
+                }
             }
         }
+
+        Ok(cryptarchia)
     }
 
-    #[expect(clippy::unused_async, reason = "To be implemented")]
-    async fn run_with<CryptarchiaState>(
+    /// Downloads a single stream of blocks from the peer,
+    /// which may or may not contain the target block.
+    /// Regardless of whether the target block is reached,
+    /// this returns the updated [`Cryptarchia`] and the last block processed.
+    /// If an error occurs during the download,
+    /// it returns [`Cryptarchia`] that contains blocks processed before the
+    /// error.
+    async fn download_stream<CryptarchiaState, BlockProcessor>(
+        &self,
         cryptarchia: Cryptarchia<CryptarchiaState>,
-        _network_adapter: NetAdapter,
-    ) -> Cryptarchia<CryptarchiaState>
+        network_adapter: &NetAdapter,
+        peer: NetAdapter::PeerId,
+        target: HeaderId,
+        latest_downloaded_block: Option<HeaderId>,
+        block_processor: &mut BlockProcessor,
+    ) -> Result<
```

Comment:
we don't need to return cryptarchia in case of an error, we can just clone the struct (which is done anyway when updating)